### ISSUE TITLE
Move Elasticsearch for RC Staging from Compose to elastic.co

### ISSUE THREAD

**Issue by adinuca** [NRGI/resourcecontracts.org#1132]
## Why
- We want to move away from Compose as their customer support is not the best and they to not have an alerting system to let you know if your deployments have scaled up. This has caused huge unexpected invoices in the past.
- Elasticsearch version currently used is no longer maintained and Compose will no longer support it starting with October 31st

## What
-  [x] Data from Compose deployment `resourcecontracts-staging` is moved from Compose to elastic.co

**Comment by adinuca** [comment]
Data for staging has been moved to elastic.co but this has caused changing in the structure.
The string fields have been converted to text fields or keyword fields according to : https://www.elastic.co/blog/strings-are-dead-long-live-strings

The following error has appeared prior to the migration from Compose to elastic.io :
```
Fatal error: Uncaught Elasticsearch\Common\Exceptions\BadRequest400Exception: {"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"Fielddata is disabled on text fields by default. Set fielddata=true on [en.country.code] in order to load fielddata in memory by uninverting the inverted index. Note that this can however use significant memory. Alternatively use a keyword field instead."}],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"query","grouped":true,"failed_shards":[{"shard":0,"index":"rc_demo_3","node":"QxU4F0rlTOWUmCcgGak5lg","reason":{"type":"illegal_argument_exception","reason":"Fielddata is disabled on text fields by default. Set fielddata=true on [en.country.code] in order to load fielddata in memory by uninverting the inverted index. Note that this can however use significant memory. Alternatively use a keyword field instead."}}]},"status":400} in /var/www/rc-api/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php:610 Stack trace: #0 in /var/www/rc-api/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php on line 610
```

There probably are additional ones that we will find during testing.

**Comment by candrei** [comment]
@adinuca as agreed in our chat there's a significant amount of effort (ie: code analysis) involved to determine whether `keyword` or `text` field is appropriate for each field. Clearly the ES automated mapping didn't do a proper job.
I propose we preserve the `string` use for now and create a task for YI to migrate the code base to the `keyword`/`text` use as part of their next contract.
For now I suspect the following procedure should work:
- create an empty index in elastic.io that would acquire the migrated data
- define its mapping based on the automatically determined schema however replace the `text` fields to `string` to match the Compose.io mapping
- import the data from Compose.io into the newly defined index

**Comment by adinuca** [comment]
Unfortunately we cannot create new mapping with type String in ES 5: https://www.elastic.co/guide/en/elasticsearch/reference/5.5/string.html

**Comment by adinuca** [comment]
It seems to work by changing the index manually for the following fields:
```
country.code
signature_year
resource
company_name
corporate_grouping
```

Will have to test it properly in staging.

**Comment by adinuca** [comment]
I've managed to get everything working, except the query to retrieve details about the contract which is used to  retrieve the PDF/text form for preview.

**Comment by adinuca** [comment]
```
Oct 25 18:57:23 772c1437cbb9 ecs-rc-api-beta-elasticsearch.log:  [2018-10-25 15:57:23] log.DEBUG: Response [{"took":1,"timed_out":false,"_shards":{"total":5,"successful":5,"skipped":0,"failed":0},"hits":{"total":0,"max_score":null,"hits":[]}}] []
Oct 25 18:57:29 772c1437cbb9 ecs-rc-api-beta-elasticsearch.log:  [2018-10-25 15:57:28] log.DEBUG: Request Body ["{\"size\":10000,\"query\":{\"bool\":{\"must\":[{\"term\":{\"open_contracting_id\":{\"value\":\"ocds-591adf-6068198767\"}}}]}}}"] []
Oct 25 18:57:29 772c1437cbb9 ecs-rc-api-beta-elasticsearch.log:  [2018-10-25 15:57:28] log.INFO: Request Success: {"method":"GET","uri":"https://54e756e310564547a3ad70b2275c5a3c.us-east-1.aws.found.io:9243/rc_demo_3/pdf_text/_search","headers":{"Host":["54e756e310564547a3ad70b2275c5a3c.us-east-1.aws.found.io:9243"],"Content-Type":["application/json"],"Accept":["application/json"]},"HTTP code":200,"duration":0.557961} []
Oct 25 18:57:29 772c1437cbb9 ecs-rc-api-beta-elasticsearch.log:  [2018-10-25 15:57:28] log.DEBUG: Response [{"took":1,"timed_out":false,"_shards":{"total":5,"successful":5,"skipped":0,"failed":0},"hits":{"total":0,"max_score":null,"hits":[]}}] []
```

**Comment by adinuca** [comment]
```
https://54e756e310564547a3ad70b2275c5a3c.us-east-1.aws.found.io:9243/rc_demo_3/pdf_text/_search
```
```
{"size":10000,"query":{"bool":{"must":[{"term":{"open_contracting_id":{"value":"ocds-591adf-6068198767"}}}]}}}
```

**Comment by candrei** [comment]
@adinuca the definition of the field has changed a bit in ES 5. In order to access the actual value (not full text indexed version of the field) you'd have to issue a query like:
```
{
  "size": 10000,
  "query": {
    "bool": {
      "must": [
        {
          "term": {
            "open_contracting_id.keyword": {
              "value": "ocds-591adf-6068198767"
            }
          }
        }
      ]
    }
  }
}
```

**Comment by adinuca** [comment]
Hi @charlesyoung , 

Staging environment is using Elasticsearch from Elastic, not Compose. I think we solved all the problems, but it would be good if you could take a look and let me know if there is anything else that still doesn't work.

Thank you!

**Comment by adinuca** [comment]
This has been done. The code will be merged to Production in #1138 .
